packages/protocol/contracts/governance/Validators.sol

```diff
@@ -35,7 +37,7 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
     string identifier;
     string name;
     string url;
-    bytes publicKey;
+    bytes publicKeysData;
```

Comment:
@martinvol, I just chatted with @kevjue about this.
Originally, I tried working with 3 separate bytes variables. This made me go over the stack size limit in Solidity, and after a few attempts this was the solution I could come up with.

Do you have any ideas for a different solution?

Reply:
Yep. We've run into this before. This seems like an ok way of dealing with this in this case. Another possibility would be to break registration into multiple steps so you don't need to supply all these variables at once. Definitely a usability/clarity trade-off between multiple calls and packing byte arrays.

Reply:
Out of curiosity, what do we gain from including the proof of possession here? It doesn't look like it's dependent on any nonce so if I understand correctly, someone could later just reuse the same public key + PoP even if they don't own it?

Reply:
Oh, I guess one other thing related to the original question: if we stay with a single function call we probably can't break up this up in the function call, butmaybe we could split up `publicKeysData` when finally storing it in a `ValidatorGroup` struct?

Reply:
> Yep. We've run into this before. This seems like an ok way of dealing with this in this case. Another possibility would be to break registration into multiple steps so you don't need to supply all these variables at once. Definitely a usability/clarity trade-off between multiple calls and packing byte arrays.

Cool idea, I think this would make this more complex that we want to and error-prone. I think I'd prefer to not do that if the benefits aren't large.



> Out of curiosity, what do we gain from including the proof of possession here? It doesn't look like it's dependent on any nonce so if I understand correctly, someone could later just reuse the same public key + PoP even if they don't own it?

The proof-of-possession prevents an attack on the BLS aggregation side, where someone who doesn't know the private key for a specific public key could inject a bad public key (e.g., a negative of the other ones), cancelling their contribution.

There is, however, a front-running or just reuse attack, where someone might copy the proof-of-possession and won't be able to sign with that key. We're working on solutions for this, though it's probably going to be adding more data to the PoP, such as the address.



> Oh, I guess one other thing related to the original question: if we stay with a single function call we probably can't break up this up in the function call, butmaybe we could split up `publicKeysData` when finally storing it in a `ValidatorGroup` struct?

Great idea. I'll give that a try.

Reply:
>I think I'd prefer to not do that if the benefits aren't large.

Yep, sounds good to me either way.

>The proof-of-possession prevents an attack on the BLS aggregation side, where someone who doesn't know the private key for a specific public key could inject a bad public key (e.g., a negative of the other ones), cancelling their contribution.

Ah cool, that makes sense.

Reply:
Doing the separate keys storage made us go over the max code size. We decided for now to leave it as it was and handle it in a longer term effort of making this contract smaller.